FEATURE: Implement strategies and add 'sync' strategy

gradinarufelix opened the pull request:
This PR adds a setting to switch between different strategies.

# Strategy "once"

Behaves like the current version of the package, translating the node *once* on adoption.

# Strategy "sync"

It translates and syncs the node according to a translation mapping setting every time the node in the source language is updated.

 # Further comments

We could actually elevate this from general setting level to a node setting level, like:

```yaml
'Vendor.Site:Document.MyPage':
  options:
    translation: 'none' | 'once' | 'sync'
```

mficzel commented:
Would love to see the strategy adjustable for each language via options.deeplStrategy in the dimension configuration.

Also the renaming of translateOnAdoption to translate is breaking. It makes sense but I would prefer a not so generic name that will likely collide with other packages. maybe automaticTranslation would do.

gradinarufelix commented:
Got it. But then I'd also suggest a more generic name here:

```
Neos:
  ContentRepository:
    contentDimensions:
      'language':
        presets:
          'de':
            label: 'German'
            values: ['de']
            uriSegment: 'de'
            options:
              deeplLanguage: 'de'
              translationStrategy: ~ | 'once' | 'sync'
              translateTo:
                - 'it'
                - 'en'
```

I'd also add a configuration checkup in `initializeObject()` in order to avoid someone creating a bad recursion.

mficzel commented:
Hmm translate to makes sense for sync but translatefrom for new translations … currently they translate from the dimension that was used to create the translation … not sure

mficzel commented:
Maybe a sane first step would be to translate from the default language on sync

gradinarufelix commented:
Thanks for the input, I already implemented it. I also added a general setting for node types `options. automaticTranslation` which is true by default and allows to turn off automatic translation for a certain node type in general. This would spare some lines of code in that case.

mficzel commented on README.md:
```diff
@@ -68,10 +79,17 @@ Sitegeist:
+      # (*) once: will translate the node once the editor switches the language dimension and only then
+      # (*) sync: will translate and sync the node anytime the base language has been edited
+      #
+      strategy: 'once'
```
Is that  still true? I thought this is configured in the contentDimensionOptions now

gradinarufelix replied:
I forgot about that, will remove it.

mficzel commented on README.md:
```diff
@@ -102,16 +137,17 @@ Neos:
             uriSegment: 'de_bar'
             options:
-              deeplLanguage: ~
+              translationStrategy: 'sync'
```
This should be `none` as deepl does not support bavarian yet

gradinarufelix replied:
Will correct that

mficzel reviewed (commented):
I would love to have `options.translateOnAdoption` still supported as an alias for `options.automaticTranslation`. As this is one line we can avoid to be breaky.

gradinarufelix commented on Classes/ContentRepository/NodeTranslationService.php:
```diff
@@ -95,7 +170,8 @@ public function afterAdoptNode(NodeInterface $node, Context $context, $recursive
             $isInlineEditable = $propertyDefinitions[$propertyName]['ui']['inlineEditable'] ?? false;
-            $isTranslateEnabled = $propertyDefinitions[$propertyName]['options']['translateOnAdoption'] ?? false;
+            // @deprecated Fallback for renamed setting translateOnAdoption -> automaticTranslation
+            $isTranslateEnabled = $propertyDefinitions[$propertyName]['options']['automaticTranslation'] ?? ($propertyDefinitions[$propertyName]['options']['translateOnAdoption'] ?? false);
```
@mficzel There is a fallback for `options.translateOnAdoption`

mficzel commented on README.md:
```diff
@@ -82,36 +93,52 @@ Sitegeist:
         presets:
-        
-          # 
+
```
Would love to see this in here to show the meaning of the `defaultPreset`

```
        #
        # The `defaultPreset` marks the source of for all tranaslations whith mode `sync`
        #  
        label: 'Language'
        default: 'en'
        defaultPreset: 'en'
        
        presets:
          
          #
          # English is the main language of the editors and spoken by editors,
          # the automatic translation is disabled therefore
          #
          'en':
            label: 'English'
            values: ['en']
            uriSegment: 'en'
            options:
              translationStrategy: 'none'

gradinarufelix replied:
Perfect!

mficzel commented:
@gradinarufelix i pushed some changes. Mainly to separate $presetIdentifiers like "en_UK from languages like "en".

gradinarufelix commented:
Adjustments look good to me. I also came across a bug and solved it: in sync mode, we also want to remove the node variants if the source node was removed. I moved this logic away from the general `translateNode` method because this is only valid for the sync mode.

mficzel commented:
@gradinarufelix adjusted one last thing. 

```
'Neos.Neos:Node':
  options:
      automaticTranslation: true
```

This will not only add this to documents and contents but also to contentCollections which imho makes sense
